### PR TITLE
Feature: Add dual printing material settings

### PR DESCRIPTION
**Pull request by bbaa77770**

### CODE REVIEW

File: resources/CuraEngine/Config/printing/snapmaker_extruder_0.def.json [hunk 1]
```diff
@@ -1,10 +1,11 @@
...
-    "name": "Extruder 1",
-    "version": 2,
+    "version": "4.1.1",
+    "name": "Extruder 2",
+    "category": "",
     "inherits": "fdmextruder",
     "metadata": {
         "machine": "Snapmaker",
-        "position": "0"
+        "position": "1"
```
Comment (xiekun1992) [inline resources/CuraEngine/Config/printing/snapmaker_extruder_0.def.json:8]: `position` should be `"0"`

File: src/app/flux/printing/index.js [hunk 2]
```diff
@@ -310,15 +330,28 @@ export const actions = {
         gcodeLineGroup.position.set(-size.x / 2, -size.y / 2, 0);
     },
 
-    updateDefaultConfigId: (type, defaultId) => (dispatch, getState) => {
+    updateDefaultConfigId: (type, defaultId, direction = 'left') => (dispatch, getState) => {
```
Comment (xiekun1992) [inline src/app/flux/printing/index.js:333]: `'left'` => `LEFT_EXTRUDER` ?

File: package.json [hunk 3]
```diff
@@ -170,6 +170,7 @@
     "@trendmicro/react-toggle-switch": "~0.5.5",
     "@trendmicro/react-validation": "~0.1.0",
     "@tweenjs/tween.js": "~17.2.0",
+    "ant-design": "^1.0.0",
```
Comment (xiekun1992) [inline package.json:173]: icons should be provided by UI designer, remove this dependency

File: src/app/ui/views/PrintingManager/PrintingManager.jsx [hunk 4]
```diff
@@ -46,6 +50,10 @@ function PrintingManager() {
     const series = useSelector(state => state?.machine?.series);
     const toolHead = useSelector(state => state?.machine?.toolHead);
     const dispatch = useDispatch();
+    const printingMaterialConfigKeys = (toolHead.printingToolhead === SINGLE_EXTRUDER_TOOLHEAD_FOR_SM2 ? PRINTING_MATERIAL_CONFIG_KEYS_SINGLE : PRINTING_MATERIAL_CONFIG_KEYS_DUAL);
```
Comment (xiekun1992) [inline src/app/ui/views/PrintingManager/PrintingManager.jsx:53]: combine these with a `if` statement

**Review by xiekun1992 — CHANGES_REQUESTED** [review]

File: src/app/ui/widgets/PrintingMaterial/Material.jsx [hunk 5]
```diff
@@ -73,14 +80,101 @@ function Material({ widgetActions }) {
...
 
+    function setDiameter(direction, value) {
+        const def = (direction === LEFT_EXTRUDER ? extruderLDefinition : extruderRDefinition);
+        def.settings.machine_nozzle_size.default_value = value;
+        dispatch(printingActions.updateExtuderDefinition(def, direction));
+    }
+    const [diametersOptions, setDiametersOptions] = useState([
+        { value: 0.2, label: 0.2 },
+        { value: 0.4, label: 0.4 },
+        { value: 0.6, label: 0.6 }
```
Comment (jane-rose) [inline src/app/ui/widgets/PrintingMaterial/Material.jsx:91]: move `constans` outside the component

File: src/app/ui/widgets/PrintingMaterial/Material.jsx [hunk 6]
```diff
@@ -106,45 +106,47 @@ function Material({ widgetActions }) {
...
     function dropdownRender(direction = LEFT_EXTRUDER) {
         return (
             (menu) => (
                 <div>
                     {menu}
                     <Divider style={{ margin: '0' }} />
-                    <div style={{ display: 'flex', flexWrap: 'nowrap' }}>
+                    <div className="height-36 padding-right-4 padding-bottom-4" style={{ display: 'flex', flexWrap: 'nowrap' }}>
                         <Input
+                            className="height-32"
```
Comment (jane-rose) [inline src/app/ui/widgets/PrintingMaterial/Material.jsx:115]: Number input?

**Review by xiekun1992 — APPROVED** [review]

**Pull request merged**